core/common/src/main/java/alluxio/network/connection/ThriftClientPool.java

```diff
@@ -0,0 +1,216 @@
+
+  /**
+   * The patterns of exception message when client and server transport frame sizes do not match
+   * or corrupted data (e.g. due to incorrect port configured).
+   */
+  private static final Pattern FRAME_SIZE_TOO_LARGE_EXCEPTION_PATTERN =
+      Pattern.compile("Frame size \\((\\d+)\\) larger than max length");
+  private static final Pattern FRAME_SIZE_NEGATIVE_EXCEPTION_PATTERN =
+      Pattern.compile("Read a negative frame size");
+
+  /**
+   * Creates a thrift client pool instance with a minimum capacity of 1.
+   *
+   * @param serviceName the service name (e.g. BlockWorkerClient)
+   * @param serviceVersion the service version
+   * @param address the server address
+   * @param maxCapacity the maximum capacity of the pool
+   * @param gcThresholdInSecs when a channel is older than this threshold and the pool's capacity
+   *        is above the minimum capaicty (1), it is closed and removed from the pool.
+   */
+  public ThriftClientPool(String serviceName, long serviceVersion, InetSocketAddress address,
+      int maxCapacity, int gcThresholdInSecs) {
+    super(Options.defaultOptions().setMaxCapacity(maxCapacity));
+    mTransportProvider = TransportProvider.Factory.create();
+    mServiceName = serviceName;
+    mServiceVersion = serviceVersion;
+    mAddress = address;
+    mGcThresholdInSecs = gcThresholdInSecs;
+  }
+
+  @Override
+  protected void closeResource(T client) {
+    // Note that the input and output protocol is the same in Alluxio.
+    TTransport transport = client.getOutputProtocol().getTransport();
+    if (transport.isOpen()) {
+      LOG.info("Closing thrift transport.");
+      transport.close();
+    } else {
+      LOG.info("Close a closed thrift transport.");
+    }
+  }
+
+  @Override
+  protected void closeResourceSync(T client) {
+    closeResource(client);
+  }
+
+  /**
+   * Creates a thrift client instance.
+   *
+   * @return the thrift client created
+   * @throws IOException if it fails to create a thrift client
+   */
+  @Override
+  protected T createNewResource() throws IOException {
+    TTransport transport = mTransportProvider.getClientTransport(mAddress);
+    TProtocol binaryProtocol = new TBinaryProtocol(transport);
+    T client = createThriftClient(new TMultiplexedProtocol(binaryProtocol, mServiceName));
+
+    RetryPolicy retry =
+        new ExponentialBackoffRetry(CONNECTION_OPEN_RETRY_BASE_SLEEP_MS, Constants.SECOND_MS,
+            CONNECTION_OPEN_RETRY_MAX);
+    while (true) {
+      try {
+        if (!transport.isOpen()) {
+          transport.open();
+        }
+        if (transport.isOpen()) {
+          checkVersion(client);
+        }
+      } catch (TTransportException e) {
+        LOG.error(
+            "Failed to connect (" + retry.getRetryCount() + ") to " + getServiceNameForLogging()
+                + " @ " + mAddress, e);
+        if (!retry.attemptRetry()) {
+          throw new IOException(e);
+        }
+      }
+      break;
+    }
```

Comment:
It is the same as Netty pool. the client doesn't need to worry about the connection after this change. I think you are suggesting to move this connection logic to RetryHandlingBlockWorkerClient. That way, we have to have a wrapper over the AlluxioService.Client. That is not very clean. I can discuss this with you in person.